### PR TITLE
Write tests for network status

### PR DESCRIPTION
**Pull request by galalen**
## Description
Write tests for [NetworkStatusDataSource](https://github.com/OpenMined/KotlinSyft/issues/127) and [NetworkStatusRepository](https://github.com/OpenMined/KotlinSyft/issues/131)
## Affected Dependencies
None.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syftlib/src/test/java/org/openmined/syft/monitor/network/NetworkStatusRealDataSourceTest.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    private val networkStatusRealTimeDataSource = mock<NetworkStatusRealTimeDataSource> {
+        on { updatePing(workerId, networkStatusModel) }.thenReturn(Completable.complete())
+        on { updateDownloadSpeed(workerId, networkStatusModel) }.thenReturn(Completable.complete())
+        on { updateUploadSpeed(workerId, networkStatusModel) }.thenReturn(Completable.complete())
+        on { getNetworkValidity(constraints) }.thenReturn(true)
+    }
+
+    @Test
+    fun `network validity should return true if network constraints satisfied`() {
+        assert(networkStatusRealTimeDataSource.getNetworkValidity(constraints))
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/monitor/network/NetworkStatusRealDataSourceTest.kt:60]: This test is bound to succeed always since in stubbing you have asked the method to always return true

Comment (vkkhare) [reply syftlib/src/test/java/org/openmined/syft/monitor/network/NetworkStatusRealDataSourceTest.kt:60]: One way to test whether the networkValidity works properly is by setting up the relevant network as active on the context. Robolectric might come on handy over here
Check the file in integration/AbstractSyftWorkerTest

**Comment by vkkhare** [comment]
The way to test NetworkDatasource would be to create mock HTTP client for downloader which would just return the expected response from server. The datasource methods will then run as expected on the mocked response.

You will have to create a mock exactly like

File: buildSrc/src/main/java/Dependencies.kt [hunk 2]
```diff
@@ -44,7 +44,7 @@ object Versions {
     const val mockitoCore = "3.2.4"
     const val mockitoKotlin = "2.2.0"
     const val junit = "4.12"
-    const val adxCore = "1.1.0"
+    const val adxCore = "1.0.0"
```
Comment (vkkhare) [inline buildSrc/src/main/java/Dependencies.kt:47]: this change should not be needed I guess

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    fun `updateDownloadSpeed return success response from server`() {
+        networkDataSourceSuccess
+                .updateDownloadSpeed(workerId, networkStatusModel)
+                .test()
+                .assertComplete()
+    }
+
+    @Test
+    fun `updateDownloadSpeed should handle error when request fail`() {
+        networkDataSourceFailure
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:101]: Better assert on the error thrown here

File: buildSrc/src/main/java/Dependencies.kt (hunk 2 above)
Comment (galalen) [reply buildSrc/src/main/java/Dependencies.kt:47]: it didn't work for me before that

Comment (vkkhare) [reply buildSrc/src/main/java/Dependencies.kt:47]: That is weird! Why would it require to downgrade the lib. It works fine for all of us. Can you send the error you are getting

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+@ExperimentalUnsignedTypes
+class NetworkStatusRealDataSourceTest : AbstractSyftWorkerTest() {
+
+    private val workerId = "test id"
+    private val networkStatusModel = NetworkStatusModel()
+
+    private var networkDataSourceSuccess: NetworkStatusRealTimeDataSource
+    private var networkDataSourceFailure: NetworkStatusRealTimeDataSource
+
+    init {
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:22]: This could go in @Before setup() check syftJobTest

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            SyftConfiguration.NetworkingClients.HTTP
+        )
+
+        networkDataSourceFailure = NetworkStatusRealTimeDataSource.initialize(configurationFailure)
+    }
+
+    @Test
+    fun `updatePing return success response from server`() {
+        networkDataSourceSuccess
+                .updatePing(workerId, networkStatusModel)
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:79]: Instead of asserting for completion, It would be better to see if a non null ping value is returned. You may require to ask the thread to wait for sometime here

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        networkDataSourceFailure
+                .updatePing(workerId, networkStatusModel)
+                .test()
+                .assertError(NetworkErrorException::class.java)
+    }
+
+    @Test
+    fun `updateDownloadSpeed return success response from server`() {
+        networkDataSourceSuccess
+                .updateDownloadSpeed(workerId, networkStatusModel)
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:95]: Same here. Check for non null downloadspeed

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+import org.junit.Test
+import org.openmined.syft.common.AbstractSyftWorkerTest
+import org.openmined.syft.common.HttpClientMock
+import org.openmined.syft.common.SocketClientMock
+import org.openmined.syft.domain.SyftConfiguration
+import org.openmined.syft.monitor.network.NetworkStatusModel
+import org.openmined.syft.monitor.network.NetworkStatusRealTimeDataSource
+
+@ExperimentalUnsignedTypes
+class NetworkStatusRealDataSourceTest : AbstractSyftWorkerTest() {
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:14]: Does this class need abstract worker for all the tests? Abstract syft worker uses robolectric which makes tests slower. If possible use it only when you absolutely require the context (it cannot be mocked)

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRepositoryTest.kt [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Test
+    fun `network validity call NetworkStatusDataSource's getNetworkValidity correctly`() {
+        val networkStatusRealDataSource = mock<NetworkStatusRealTimeDataSource> {
+            on { getNetworkValidity(listOf()) }.thenReturn(true)
+        }
+        val networkStatusRepository = NetworkStatusRepository(listOf(), mock(), networkStatusRealDataSource)
+
+        assert(networkStatusRepository.getNetworkValidity())
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRepositoryTest.kt:46]: Instead of assert here better call verify on the mocked network status datasource

**Comment by vkkhare** [comment]
Also it seems the coverage increase is minimal here. Look for files with specifically low coverage amongst and add tests for the red line in https://codecov.io/gh/OpenMined/KotlinSyft/src/86adefc331cb8dfefc3280ed03122846bf01898e/syftlib/src/main/java/org/openmined/syft/monitor/network/NetworkStatusDataSource.kt

Like datasource needs a test to verify when downloadspeeds are slower. You would need to explicitly wait on the thread for checking download speed. Don't worry too much over the classes with more than 95% coverage

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt (hunk 4 above)
Comment (galalen) [reply syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:22]: Yeah, I saw it

File: buildSrc/src/main/java/Dependencies.kt (hunk 2 above)
Comment (galalen) [reply buildSrc/src/main/java/Dependencies.kt:47]: Could not find core.aar (androidx.test:core:1.1.0).
Searched in the following locations:
    https://dl.google.com/dl/android/maven2/androidx/test/core/1.1.0/core-1.1.0.aar

Comment (vkkhare) [reply buildSrc/src/main/java/Dependencies.kt:47]: this link works though

Comment (vkkhare) [reply buildSrc/src/main/java/Dependencies.kt:47]: try invalidating the gradle cache and restart ;p

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt (hunk 7 above)
Comment (galalen) [reply syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:14]: The context is required in initializing NetworkStatusRealTimeDataSource

Comment (vkkhare) [reply syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkStatusRealDataSourceTest.kt:14]: Yeah but a mock Context can be supplied there unless we need interactions with context that can't be stubbed like setting up network validity constraints (so except for network validity I don't see the need of context and it can simply be mocked)

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    @Test
+    fun `network validity should return true if network constraints satisfied`() {
+        val networkDataSource = NetworkStatusRealTimeDataSource.initialize(config)
+        assert(networkDataSource.getNetworkValidity(networkConstraints))
+    }
+
+
+    // Test for NetworkRepository's getNetworkValidity
+    @Test
+    fun `network validity call NetworkStatusDataSource's getNetworkValidity correctly`() {
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt:66]: This could be done by calling verify over mocked data source

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+        assert(cachedStatus.ping != null)
+        assert(cachedStatus.downloadSpeed != null)
+        assert(cachedStatus.uploadspeed != null)
+        assert(cachedStatus.networkValidity)
+    }
+
+    // Test for NetworkRepository's getNetworkStatus
+    @Test
+    fun `getNetworkStatus calculate new NetworkModelStatus if there no cache`() {
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt:92]: This test can also be done via mocked instance

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    @Test
+    fun `network validity call NetworkStatusDataSource's getNetworkValidity correctly`() {
+        val networkStatusRealDataSource = NetworkStatusRealTimeDataSource.initialize(config)
+        val networkStatusRepository = NetworkStatusRepository(listOf(), mock(), networkStatusRealDataSource)
+        assert(networkStatusRepository.getNetworkValidity())
+    }
+
+    // Test for NetworkRepository's getNetworkStatus
+    @Test
+    fun `getNetworkStatus should return cached network status model`() {
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt:74]: Same here. Should not need robolectric

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt (hunk 10 above)
Comment (galalen) [reply syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt:92]: I managed to get mock the others without robolectric, but for this when calculating new model status it requires a context to get networkVaildity if I mock this function I will get null pointer error on the downloader. I'll push a draft RP

**Comment by galalen** [comment]
> Check the build failure. Rest I guess you would need to see how to delay ping/downloadspeed response. That should finish testing

Why should I delay them?

**Comment by vkkhare** [comment]
So that you can get non zero null values equal to the delay. The ping in that case should ideally be equal to the delay you added

**Comment by vkkhare** [comment]
@galalen a rebase is also needed here

**Comment by vkkhare** [comment]
We are finally getting close lol! 
Check 
https://codecov.io/gh/OpenMined/KotlinSyft/pull/208/src/syftlib/src/main/java/org/openmined/syft/monitor/network/NetworkStatusDataSource.kt

`getNetworkValidity` also needs to be tested for if constraints are returned null due to some error it throws the error properly.

Relatimedatasource needs a test for `readNbuffers` check the coverage for lines that still need testing

**Comment by galalen** [comment]
> We are finally getting close lol!
> Check
> https://codecov.io/gh/OpenMined/KotlinSyft/pull/208/src/syftlib/src/main/java/org/openmined/syft/monitor/network/NetworkStatusDataSource.kt
> 
> `getNetworkValidity` also needs to be tested for if constraints are returned null due to some error it throws the error properly.
> 
> Relatimedatasource needs a test for `readNbuffers` check the coverage for lines that still need testing

Longest PR ever lol!

**Comment by vkkhare** [comment]
It's because that portion doesn't run unless download speed is slow. We have explicitly make downloading slow to test that

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/client/DelayedHttpClientMock.kt [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+                    Response.error(
+                        403,
+                        SpeedCheckResponse("error message").toString().toResponseBody()
+                    )
+                )
+            )
+
+            on { uploadSpeedTest(any(), any(), any(), check {}) }.thenReturn(
+                if (uploadSuccess)
+                    Single.just(Response.success(SpeedCheckResponse()))
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/client/DelayedHttpClientMock.kt:48]: Delay this as well

File: syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        val networkStatusModel = NetworkStatusModel("1", "1", "1", true)
+        val cache = mock<NetworkStatusCache> {
+            on { networkStateCache }.thenReturn(networkStatusModel)
+            on { getNetworkStatusCache() }.thenReturn(Maybe.just(networkStatusModel))
+        }
+
+        val networkStatusRepository = NetworkStatusRepository(listOf(), cache, networkDataSource)
+        val cachedStatus = networkStatusRepository.getNetworkStatus("worker id", true).blockingGet()
+
+        assert(cachedStatus.ping != null)
```
Comment (vkkhare) [inline syftlib/src/test/java/org/openmined/syft/unit/monitor/network/NetworkValidityTest.kt:95]: These should test for value rather than not being null

**Comment by vkkhare** [comment]
@all-contributors add @galalen for testing
